### ISSUE TITLE
Support for Vacuum cleaners?

### ISSUE THREAD

**Issue by Claeysson** [resoai/TileBoard#138]
Is it possible to have a tile that support robot vacuumers? I've tried the standard switch but it doesn't work. Sems the vacuum type needs "start"/"stop" and not "turn_on"/"turn_off". Is it possible to configure this?

**Comment by resoai** [comment]
Please post the state dump for your vacuum cleaner

**Comment by mouth4war** [comment]
Yes, there have been changes in HASS vacuum component state management. The
switch tile no longer works for me also.

https://github.com/home-assistant/architecture/issues/29

**Comment by resoai** [comment]
Guys, I need to see full state dump (from device tools) to make a vacuum tile.

**Comment by Claeysson** [comment]
How can I generate this dump?

**Comment by resoai** [comment]
Just to to dev tools in HA and scroll down to the vacuum entity

**Comment by Claeysson** [comment]
Is this what you want?

```
battery_level: 100
battery_icon: mdi:battery-charging-100
status: Docked
clean_start: 2018-09-13 13:17:07
clean_stop: 2018-09-13 14:04:55
clean_area: 46.4752
clean_suspension_count: 0
clean_suspension_time: 0
battery_level_at_clean_start: 97
battery_level_at_clean_end: 56
friendly_name: Lisa-Marie
supported_features: 14940
```

**Comment by mouth4war** [comment]
`battery_level: 100
battery_icon: mdi:battery-charging-100
fan_speed: Max
fan_speed_list: Quiet,Balanced,Turbo,Max
do_not_disturb: on
do_not_disturb_start: 22:00:00
do_not_disturb_end: 08:00:00
cleaning_time: 41
cleaned_area: 22
cleaning_count: 758
total_cleaned_area: 12635
total_cleaning_time: 21131
main_brush_left: 246
side_brush_left: 146
filter_left: 51
sensor_dirty_left: 26
status: Charging
friendly_name: Rosie
supported_features: 14204`

**Comment by resoai** [comment]
According to the description vacuum component has both `turn_on` / `turn_off` and `start` / `stop`.

https://www.home-assistant.io/components/vacuum/

So it is a bit tricky..

**Comment by resoai** [comment]
What does the `state` show when it is `on` and `off`?

**Comment by resoai** [comment]
We need to know what other vacuums are reporting as state to actually do the tile. I've already created base support for vacuums and can decode `supported_features` i.e. know if vacuum supports `turn_on` or not.

**Comment by mouth4war** [comment]
Xiaomi doesnt work with start_pause service. Start/Turn_on and Pause/Turn_off work. Return to home also works.

State and status pairs:
```
status: Cleaning, state: cleaning
status: Paused, state: paused
status: Returning home, state: returning
supported_features: 14204
```

**Comment by resoai** [comment]
I'm a bit confused. So, to cover all vacuums toggling should act as follows:

1. If `state` is `idle` or `off` then
    if vacuum supports `turn_on` then
        send `turn_on`
    else
        send `start`

2. If `state` is `cleaning` or `on` then
    if vacuum supports `turn_off` then
        send `turn_off`
    else
        send `stop`

Is this correct?

**Comment by resoai** [comment]
This is definitely not the way to do it. We need to know exactly how each vacuum cleaner behaves at which states, we already have a capability of checking supported features which I added in the last update.

**Comment by Claeysson** [comment]
You're right. I've made a first attempt on solving the issue:

Added to index.html:
```
<div ng-if="item.type === TYPES.VACUUM"
      class="item-entity-container">
      
      <div class="item-entity">
        <span class="item-entity--icon mdi"
        ng-class="entityIcon(item, entity)"></span>
      </div>
      <div class="item-vacuum-battery">
        <i class="mdi mdi-battery"></i> <span ng-bind="getVacuumBattery(item, entity)">95%</span>
      </div>
      
</div>
```

added to main.js:


`case TYPES.VACUUM: return $scope.callVacuum(item, entity);`


```
$scope.getVacuumBattery = function (item, entity) {

     var battery = entity.attributes.battery_level + "%";
     return battery;
   };
```


```
$scope.callVacuum = function (item, entity) {
     //console.log(entity.state);
     if(entity.state === "idle" || entity.state === "docked" || entity.state === "paused") service = "start";
     else if(entity.state === "cleaning") service = "pause";
      sendItemData(item, {
         type: "call_service",
         domain: "vacuum",
         service: service,
         service_data: {
            entity_id: item.id
         }
      });
   };
```

added to main.less:

`@VACUUM: rgb(122, 78, 127);`

`&.-th-vacuum {background-color: @SWITCH;} `

`&.-vacuum,`

```
.item-vacuum-battery {  //Tillägg
   position: absolute;
   text-align: left;
   padding: 0 7px;
   bottom: 26px;
   left: 0;
   right: 0;
   font-size: 0.85em;
   font-weight: 400;
   opacity: 0.9;
   overflow: hidden;
   text-overflow: ellipsis;
   z-index: 2;
}
```

then in my config.json:
```
{
      position: [0, 0],
      type: TYPES.VACUUM,
      id: 'vacuum.lisamarie',
      title: 'Dammsugare',
      icon: 'mdi-robot-vacuum-variant',
      states: {
           paused: "Pausad",
           idle: "Vilar",
           docked: "Dockad",
           cleaning: "Städar"
       }
}
```

Works on my vacuum. Since I'm not at professional developer it might not be the cleanest solution. Whats your opinion?

**Comment by mouth4war** [comment]
Agreed that the buttons displayed (and services behind them) depend on supported features.

For starters, all the supported buttons can be shown (regardless of current state).

Later, logic to show only relevant buttons can be added (even HASS UI currently doesn't do this).

From the component, valid States are:
STATE_CLEANING = 'cleaning'
STATE_DOCKED = 'docked'
STATE_IDLE = STATE_IDLE
STATE_PAUSED = STATE_PAUSED
STATE_RETURNING = 'returning'
STATE_ERROR = 'error'

**Comment by resoai** [comment]
I've already added features: 

https://github.com/resoai/TileBoard/blob/master/scripts/app.js#L82

States can differ as far as I can see.

**Comment by mouth4war** [comment]
@Claeysson why don't you submit your changes as a PR and people can improve on it. I think your version will be good for a small tile and there can be another version with more buttons.

One change that should be made for your version is a check for features in $scope.callVacuum. Maybe a new $scope.shouldShowStartPauseVacuum to help decide which service to call (start/pause or turn_on/turn_off).

**Comment by resoai** [comment]
Please don't. I will create a basic tile when I have a bit more free time and we can extend it in future.

**Comment by Claeysson** [comment]
@resoai I think you're being counterproductive when you tell me to don't contribute to the project. Maybe my tile isn't a perfect or good solution, but please give me some feedback instead of just rejecting it.

**Comment by resoai** [comment]
I am not discouraging anyone but there are certain things to consider when doing the tile for future extendability. I have very little time but I have been adding tiles to support stuff that I don't even have whenever people asked for it.

**Comment by resoai** [comment]
Please see if your vacuum behaves as expected with latest update.
